jdaviz/configs/imviz/helper.py

```diff
@@ -209,6 +209,80 @@ def offset_by(self, dx, dy):
                 viewer.state.x_max = viewer.state.x_min + width
                 viewer.state.y_max = viewer.state.y_min + height
 
+    @property
+    def zoom_level(self):
+        """Zoom level:
+
+        * 1 means real-pixel-size.
+        * 2 means zoomed in by a factor of 2.
+        * 0.5 means zoomed out by a factor of 2.
+        * 'fit' means zoomed to fit the whole image width into display.
+
+        """
+        viewer = self.app.get_viewer("viewer-1")
+        if viewer.shape is None:  # pragma: no cover
+            raise ValueError('Viewer is still loading, try again later')
+
+        screenx = viewer.shape[1]
+        screeny = viewer.shape[0]
+        zoom_x = screenx / (viewer.state.x_max - viewer.state.x_min)
+        zoom_y = screeny / (viewer.state.y_max - viewer.state.y_min)
+
+        return max(zoom_x, zoom_y)  # Similar to Ginga get_scale()
+
+    # Loosely based on glue/viewers/image/state.py
+    @zoom_level.setter
+    def zoom_level(self, val):
+        if ((not isinstance(val, (int, float)) and val != 'fit') or
+                (isinstance(val, (int, float)) and val <= 0)):
+            raise ValueError(f'Unsupported zoom level: {val}')
+
+        viewer = self.app.get_viewer("viewer-1")
+        image = viewer.state.reference_data
+        if (image is None or viewer.shape is None or
+                viewer.state.x_att is None or viewer.state.y_att is None):  # pragma: no cover
+            return
+
+        # Zoom on X and Y will auto-adjust.
+        if val == 'fit':
+            # Similar to ImageViewerState.reset_limits() in Glue.
+            new_x_min = 0
+            new_x_max = image.shape[viewer.state.x_att.axis]
+        else:
+            cur_xcen = (viewer.state.x_min + viewer.state.x_max) * 0.5
+            new_dx = viewer.shape[1] * 0.5 / val
+            new_x_min = cur_xcen - new_dx
+            new_x_max = cur_xcen + new_dx
+
+        with delay_callback(viewer.state, 'x_min', 'x_max'):
+            viewer.state.x_min = new_x_min - 0.5
+            viewer.state.x_max = new_x_max - 0.5
+
+        # We need to adjust the limits in here to avoid triggering all
+        # the update events then changing the limits again.
+        viewer.state._adjust_limits_aspect()
+
+    # Discussion on why we need two different ways to set zoom at
+    # https://github.com/astropy/astrowidgets/issues/144
+    def zoom(self, val):
+        """Zoom in or out by the given factor.
+
+        Parameters
+        ----------
+        val : int or float
+            The zoom level to zoom the image.
+            See `zoom_level`.
+
+        Raises
+        ------
+        ValueError
+            Invalid zoom factor.
+
+        """
+        if not isinstance(val, (int, float)):
+            raise ValueError(f"zoom only accepts int or float but got '{val}'")
+        self.zoom_level = self.zoom_level * val
```

Comment:
If this is simply a method to set the `zoom_level`, it would would be nice to accept "fit" as input here as well as when setting `zoom_level` directly. That was one of the first things I tried and I was surprised that it didn't work.

Reply:
That would go against the current `astrowidgets` API upstream though...

Reply:
`zoom()` was supposed to be *relative* zoom, so "fit" does not really fit (pun?) in that paradigm.

Reply:
Ah I see, that's the context I was missing, I guess. I didn't fully appreciate that `zoom_level` is setting the absolute zoom relative to the image, whereas `zoom` is setting the zoom relative to the current display.

Reply:
I can't blame you. I didn't either until astropy/astrowidgets#144